### PR TITLE
Use `Target` as type for `pow_limit`

### PR DESCRIPTION
**Pull request by tcharding**
The `Params::pow_limit` field is currently a `Work` type, this is incorrect. The proof of work limit is the highest _target_ not the lowest work (even though these have a relationship).
    
Note that we use the highest _attainable_ target, this differs from Bitcoin Core and the reasoning is already documented in the code.
    
Add new consts and document where they came from as well as how they differ to Core.
    
Use the new consts in the various network specific `Params` types.

Fix: #2106

### CODE REVIEW

**Comment by tcharding** [comment]
Mad props to @junderw for finding these bugs, legend.

**Comment by tcharding** [comment]
FTR I did not work out where I got the `inverse` function algo from. I'm going to go ahead and trust Tobin from the past, this PR does _not_ do the changes suggested by @junderw in #2103.

File: bitcoin/src/consensus/params.rs
Comment (junderw) [inline bitcoin/src/consensus/params.rs:91]: https://github.com/bitcoin/bitcoin/blob/8105bce5b384c72cf08b25b7c5343622754e7337/src/kernel/chainparams.cpp#L348C40-L348C104

**Comment by junderw** [comment]
> FTR I did not work out where I got the `inverse` function algo from. I'm going to go ahead and trust Tobin from the past, this PR does _not_ do the changes suggested by @junderw in #2103.

https://github.com/bitcoin/bitcoin/blob/54bdb6e0745934ad9ae6c77628f2382a83b9a1f0/src/chain.cpp#L139-L143

Found it

**Comment by junderw** [comment]
I'm still not sure why we should deviate from Bitcoin Core?

Bitcoin Core has `00000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffff`

So why are we attempting to pre-lossify it into the compact target's representation? To create round trip tests?

I'd rather we do a round trip and add the extra `ff` after the roundtrip. It still makes it clear.

A round trip with a lossy operation shouldn't really be possible IMO.

**Comment by tcharding** [comment]
> Found it

Ah nice one!

> I'm still not sure why we should deviate from Bitcoin Core?

This was my thought as well as I drove away from computer after work yesterday. I came back and had a brief go at changing it and a whole bunch of unit tests failed in other places.

I also played with modifying `to_work` and `to_target` to make them saturate at MAX/MIN but I couldn't get that to work.

Perhaps @apoelstra can weigh in with some history as to why we deviate. Also it'd be nice if someone actually understood the math (I read "multiplicative inverse multiplied by some constant" and got lost).

Also, I don't know how "inverse" ties in to the Core code, Core doesn't mention "inverse" in the function name or comments.

> We need to compute 2**256 / (bnTarget+1) 

(ref `GetBlockProof` in core (https://github.com/bitcoin/bitcoin/blob/54bdb6e0745934ad9ae6c77628f2382a83b9a1f0/src/chain.cpp#L139-L143)

`GetBlockProof` does not mention "inverse"?

Another idea is to roundtrip through compact form in  `to_work` and `to_target` but this means we break the `inverse` rountripability for arbitrary values. We implement `Sub` for `Work` so arbitrary values are possible. Maybe we should enforce MIX/MAX as invariants but again this is super ugly because max is network dependant.

Perhaps I'm overthinking it?

File: bitcoin/src/consensus/params.rs
Comment (tcharding) [reply bitcoin/src/consensus/params.rs:91]: Awesome, while writing a response I work it out, we just run the Core version through `Target::from_compact(t.to_compact_lossy())` like I did for the regtest one already, deeeer I'm slow sometimes :)

Added a `Target::SIGNET_MAX` as well as permalinks on all the consts to Core.

**Comment by apoelstra** [comment]
> So why are we attempting to pre-lossify it into the compact target's representation? To create round trip tests?

This was an explicit decision made in https://github.com/rust-bitcoin/rust-bitcoin/pull/464

**Comment by apoelstra** [comment]
To be more specific: this constant serves two purposes. To be a limiting value, and to be the difficulty value for the genesis block. Using Core's constant means that the genesis block can't actually be validated because its compact difficulty representation doesn't match its actual difficulty. So instead we should use the actual difficulty.

As for the *limit*, it is purely a philosophical question as to whether we use Core's value or our value, since there are no expressible difficulties in between the two.

**Comment by junderw** [comment]
> Note that, in Bitcoin Core, the original value is only ever used once in its original form:

From that PR.

Ironically, that "once" it is used is literally the code we are trying to reproduce in #2090 lol

Maybe we should split into two constants?

**Comment by apoelstra** [comment]
Yeah, I think we should split it into two constants. Especially since Matt's long comment appears in the definition of the genesis block, but not next to the definition of the constant.

**Comment by junderw** [comment]
What about this?

```rust
pub pow_limit: Target,         // Current
pub pow_limit_precise: Target, // Precise value (from Core)
```

**Comment by apoelstra** [comment]
@junderw I think one should be `pow_limit` (the "precise" one) and the other should be `genesis_target` (the "imprecise" one).

**Comment by junderw** [comment]
Ok, I updated my comment to reflect that. Sounds good to me.

**Comment by apoelstra** [comment]
cc @TheBlueMatt since this discussion is based around your #464

**Comment by junderw** [comment]
Instead of inflating the header's bits into a Target, why not perform `required_target.to_compact_lossy()` on the required_target? Then you could do `if self.bits != required_target.to_compact_lossy() {`

https://github.com/rust-bitcoin/rust-bitcoin/blob/eab9f89779eb4b84ea8bdb5140ddc52906382250/bitcoin/src/blockdata/block.rs#L79-L82

**Comment by apoelstra** [comment]
@junderw I think that's unintuitive for users. The compact encoding is "just an encoding type" and it feels weird to do any actual computations on it, except as an optimization.

**Comment by apoelstra** [comment]
I also think that doing this inside `validate_pow` might actually be incorrect. Though I'm unsure.

**Comment by junderw** [comment]
> I'm not entirely clear on why we need an inaccurate anything at all.

In that case would you agree to reverting the inaccurate values introduced in your PR at #464 ?

Or are you referring to some other inaccuracy?

**Comment by junderw** [comment]
Currently this PR's discussion is in a state of:

- We should change these to reflect Core.
- But we changed them to NOT reflect Core for a specific reason.
- What was that reason?
- The reasoning was given in 464 where it was introduced, so let's ask @TheBlueMatt about reverting it (this PR)
- @TheBlueMatt : "I'm not entirely clear on why we need an inaccurate anything at all."


If we interpret that to "reverting 464 is OK" then I guess we can take that as a concept ACK for this PR (which is essentially reverting 464)

OR

If that is not what it's referring to, we should discuss further.

File: bitcoin/src/consensus/params.rs
Comment (junderw) [inline bitcoin/src/consensus/params.rs:92]:
```suggestion
                pow_target_spacing: 10 * 60,            // 10 minutes.
```

**Comment by tcharding** [comment]
I have limited this PR in scope to just fixing #2106, this is a bug fix and needs to be in the upcoming release.

All the other work/target min/max discussion is part of, as far as I understand it, being able to do #2090, which is a new feature and should go in the next release.

**Comment by tcharding** [comment]
Createt #2021 to track the whole pow mess. Again, mad props to @junderw for uncovering all this.

File: bitcoin/src/consensus/params.rs
Comment (tcharding) [reply bitcoin/src/consensus/params.rs]: Reminds me of one of the more painful review comments I once got early in my days as an open source hacker, "does this even compile?"

**Comment by tcharding** [comment]
Because the value of `MAX_ATTAINABLE_REGTEST` regtest is not immediately obvious I just checked this
```
assert_eq!(U256(0x7fffff00000000000000000000000000u128, 0), Target::MAX_ATTAINABLE_REGTEST.0)
``` 
Based on the value in #464

**Review by junderw — APPROVED** [review]
ACK 38005f6aa7a580e9f4d28cb1b37b10375a304e6d

In regards to somehow including Core's constant, that can be discussed in another issue.

**Comment by tcharding** [comment]
Added a link to your comment to #2121 @junderw

**Comment by tcharding** [comment]
FTR this PR has been totally re-written and is ready for review please crew.

**Comment by apoelstra** [comment]
The current version looks great!

**Review by apoelstra — APPROVED** [review]
ACK 38005f6aa7a580e9f4d28cb1b37b10375a304e6d

**Comment by tcharding** [comment]
It is going to be hard to get a second ack from a maintainer on this, this work/target stuff requires thought so no one is going to sign off on it without putting in time to think about it. In my opinion the ack from @junderw counts because he has already put in the time and effort to think about this stuff.

Also this is a bug fix and is holding up the RC release, IMO this should merge.

**Comment by apoelstra** [comment]
I agree, the ACK from @junderw should definitely count. Thanks for poking me.

File: bitcoin/src/pow.rs
Comment (sanket1729) [inline bitcoin/src/pow.rs:122]: Why is the text for doc-comments different for mainnet and others. Should we not have uniform doc-comment across of all them? Can we also add some constants

**Review by sanket1729 — COMMENTED** [review]
Thanks @tcharding, I was about to comment that this one is hard for me for review especially because I cannot directly compare constants with core. Are there any test cases that we can write to reproduce the values you used in the constants?

File: bitcoin/src/pow.rs
Comment (tcharding) [reply bitcoin/src/pow.rs:122]: Oh, that's sloppy work by me. They should all be uniform. I added a note to #2107. At one stage I had all the constants from core with tests that when converted to compact form they matched these but I dropped it all to get this bug fix diff as small as possible.

**Comment by apoelstra** [comment]
@sanket1729 regarding test cases, we could start with Core's constant, convert it to a compact target lossily, then re-expand it and assert that the result equals our constant.

**Pull request merged**
